Jsx table

Lamarr20 opened the pull request:
This pull request adds a table component inside of src/components/04-tables. It takes a config object and with two functions transform an array of strings into its appropriate table format. The `createHeader` function creates a dynamic column header and can be composed of 1 or more headers. Then `createRow` function takes an array of strings and separates each array to match its corresponding row depending on the order of the array. 

Closes: nyc-cto/USWDS-generator/issues/60

Lamarr20 committed:
table generator

Lamarr20 committed:
add id tag

Lamarr20 committed:
break line at <tr> and </tr>

Lamarr20 committed:
fix typo

Lamarr20 committed:
add jsdoc documentation

MathewDavidov commented:
Need to update JSDoc - can use [this as reference](https://github.com/nyc-cto/uswds/blob/589ad2461208bffa83cf6dc75251c24df3bfdcff/src/components/02-buttons/button.jsx)

Lamarr20 committed:
Update jsdoc and table component

MathewDavidov commented:
> Need to update JSDoc - can use [this as reference](https://github.com/nyc-cto/uswds/blob/589ad2461208bffa83cf6dc75251c24df3bfdcff/src/components/02-buttons/button.jsx)

[In this commit](https://github.com/nyc-cto/uswds/pull/11/commits/d696f0f500c8d6bb6e0a4fc454483e3f42146707), Lamarr and I fixed JSDoc formatting, tags, descriptions, and names. Additionally, we reworked the functions to be compliant with the rules and be consistent and cleaner.

JohnAKASquib committed:
Merge branch 'jsxTable' of https://github.com/JohnAKASquib/uswds into jsxTable

JohnAKASquib committed:
Inner parts of table now have their own functions
Table row and header have their own function to return HTML,
Main Table had its main body replaced with config.children

JohnAKASquib commented:
Closes [nyc-cto/USWDS-generator/issues/70](https://github.com/nyc-cto/USWDS-generator/issues/70)

jisaf commented on src/components/04-tables/tables.jsx:
```diff
@@ -0,0 +1,96 @@
+  return (
+    <table id={`${config.tableID}`} class={`usa-table ${config.class}`}>
+      <caption>{config.caption}</caption>
+      {config.children}
```
`children` may need to be changed to be more generic and/or accommodate other frameworks, but it should work for now.

jisaf commented on src/components/04-tables/tables.jsx:
```diff
@@ -0,0 +1,96 @@
+ */
+exports.Tables = function (config) {
+  return (
+    <table id={`${config.tableID}`} class={`usa-table ${config.class}`}>
```
non-blocker, but is there a way to infer or generate an id so that it doesn't need to be passed in? are there problems that arise if it is inferred or generated?

JohnAKASquib replied:
There's probably a random number generator JavaScript has that I could use to create one.

jisaf commented on src/components/04-tables/tables.jsx:
```diff
@@ -0,0 +1,96 @@
+  let tableHeader = [];
+
+  for (let i = 0; i < config.header.length; i++) {
+    tableHeader.push(`<th scope='col'>${config.header[i]}</th>`);
```
non-blocker: is there a use case where a developer might want to pass a class to a `<th>`? If so, it might be useful to add make each header an object: `<th ${config.header[i].class} scope='col'>${config.header[i].text}</th>`. However there are complexity downsides to this which would need to be weighed

JohnAKASquib replied:
I took your advice and did exactly this. Each header is an object with **class** and **text** properties.

jisaf commented on src/components/04-tables/tables.jsx:
```diff
@@ -0,0 +1,96 @@
+
+  return (
+    <thead>
+      <tr>{tableHeader.join("\n")}</tr>
```
non-blocker: In React and JSX there is a common pattern that you will see that uses `.map()` to replace the for loop and would be used inline here:

```
{config.header.map(header => (
  <th scope='col'>{header}</th>
))}
```

This uses the ES6 `.map()` function though which may look abnormal so developers not familiar with it, so there is a trade-off discussion to have about which one is more maintainable.

jisaf commented on src/components/04-tables/tables.jsx:
```diff
@@ -0,0 +1,96 @@
+ * TableRow(config);
+ * @returns {HTMLTableRowElement} - The HTML for the table row
+ */
+exports.TableRow = function (config) {
```
Consider either making this a single TableRow (i.e. no `for` loop, and let the developer to that part) or name it plural TableRows. Naming conventions generally should follow the input and/or output of the function.

As a side note, and definitely not a blocker, I would generally lean towards making this a Singular function and then as a user of the React version I would implement it as:

```
<Table name="Table" id="demo">
  <TableHeader header=["foo", "bar", "baz"]/>
  {tableRows.map(row => <TableRow row={row.data}/>)}
</Table>
```

In this case my tableRows would look like:
```
tableRows = [
  ["foo", "bar", "baz"],
  ["oof", "rab", "zab"]
]
```

This would let you simplify your code, while making it a bit more clear to the end dev what is going on.

JohnAKASquib replied:
I changed the **TableRow** function so that it only returns 1 `<tr>` element so its not confusing with its name. I also made it so a class can be passed into `<th>` just like you mentioned with the **TableHeader**.

JohnAKASquib committed:
TableRow function now returns only 1 <tr> element

JohnAKASquib committed:
TableRow and TableHeader allow for classes to be passed in

JohnAKASquib committed:
fixed some mistakes in the JSDoc
